### PR TITLE
Asyncified validate + submit flow

### LINKED ISSUE #644: Validation Errors should not rely on throw/catch
Validators currently work by throwing an object that is at a minimum `{ details: arrayOfErrors}`

This is not best practice as validation errors are not actually runtime errors but the proper result of a working validator against a currently invalid model.

In future versions it would be better to handle this via returning errors directly (versus returning nothing to indicate success), using promises, error first callbacks, etc.

**Side note:** 

The current system also requires throwing a somewhat nonstandard Error object that cannot be created via the Error() constructor alone.  If I am using, for instance the JSONSchemaBridge and want to throw a standard Error object (as to appease a linter) I would need the following code:

```
throw Object.assign(new Error('Validation Error'), { details: validator.errors })
```

Of course one could argue throwing a standard Error object is perhaps worse because, once again, this is not an actual error condition.  In my case I settle on just disabling ESLint for the line in question.

### PR DESCRIPTION
**Pull request by radekmie**
Let's start with a short flashback from #644:
> [...] Well, the first bridge ever was the SimpleSchema (the Meteor one), and its validator throws an error instead of returning it. It was perfectly fine, up to the point when we wanted to add `onSubmitSuccess`/`onSubmitFailure` and async validation. Of course, we managed to do so, but the solution is not very elegant.

This _not very elegant solution_ got implemented in #51. Why it's not elegant? Because it's redundant. There's no need for such hooks if one can simply add it directly to the `onSubmit`. Another problem (non-elegancy) is the asynchronous validation - invented in #17 and implemented in 4a56596f8a2e296f891316430cdba104cd21f8e6. It's weird and is using callbacks.

This PR introduces a new, clean flow (including https://github.com/vazco/uniforms/pull/711#issuecomment-618319641):
* **Bridge validators can to be asynchronous.**
* **Bridge validators have to return errors instead of throwing them.** More background is in #644.
* **The `onValidate` is now `(model, error?) => error? | Promise<error?>`.** This is far better and obvious than the callback version before.
* **Removed `onSubmitSuccess` and `onSubmitFailure`.** I'm pretty sure there's no viable case for them at all. Migration is straightforward:
  ```diff
  -onSubmit={onSubmit}
  -onSubmitSuccess={onSubmitSuccess}
  -onSubmitFailure={onSubmitFailure}
  +onSubmit={model => {
  +  const result = onSubmit(model);
  +  result.then(onSubmitSuccess, onSubmitFailure);
  +  return result;
  +}}`
  ```
* **Separated `submitting` and `validating` states.** Now `validating` is `true` only while the bridge validator or `onValidate` are currently running. Same for `submitting` - it's try only while `onSubmit` is running.

**TODO:**
- [x] Perform some manual tests to check if there's no visible difference with this change. It might be the case as few things got _asyncified_.
- [x] Run benchmarks to ensure no performance degradation.
- [x] Add missing tests.
- [x] Update `CHANGELOG.md`.

---

Closes #644.

### CODE REVIEW

**Comment by radekmie** [comment]
During benchmarks, it turned out that making everything async has one flaw: `setState` happens before `then`. It means that we have at least one additional render. Right now I'm working on a solution similar to [the one in `BaseForm`](https://github.com/vazco/uniforms/blob/04e1ff427998f1369f39945ad416f443c2c0ff7a/packages/uniforms/src/BaseForm.tsx#L281-L288) to handle both sync and async validators, `onValidate`, and `onSubmit` in any configuration.

Another problem araised while working with rejected promises. The thing is that it's not better than `throw` as we still cannot type it anyhow. We won't be able to type it correctly anytime soon (I don't have an idea on how to type errors as those are created in bridges, used in forms, and then analyzed by bridges), but having such a possibility would be nice. I'll experiment with actually returning errors instead, as discussed in #644.

File: packages/uniforms/__tests__/ValidatedForm.tsx [hunk 1]
```diff
@@ -389,4 +341,122 @@ describe('ValidatedForm', () => {
...
+      return Object.keys(x) as (keyof X)[];
+    }
+
+    const cases = cartesian(
+      [true, false] as [true, false],
+      cartesian(
+        keys(variantGroups[0]),
+        cartesian(keys(variantGroups[1]), keys(variantGroups[2])),
+      ),
+    ).map(([a, [b, [c, d]]]) => [a, b, c, d] as const);
```
Comment (kestarumper) [inline packages/uniforms/__tests__/ValidatedForm.tsx:391]: This is clever, but unreadable. I don't think anyone in the future will know what is going on here (except maybe the author). But let's leave it as is.

File: packages/uniforms/src/ValidatedForm.tsx [hunk 2]
```diff
@@ -132,34 +121,24 @@ export function Validated<Base extends typeof BaseForm>(Base: Base) {
...
-            // If validation fails, or `super.onSubmit` doesn't touch `submitting`, we need to reset it.
-            this.setState(state =>
-              state.submitting ? { submitting: false } : null,
-            );
-          }
+        // Validation failed (i.e. returned an error), so no error is present
+        // both in the props nor the state.
+        return super.onSubmit().catch(error => {
+          this.setState({ error });
+          throw error;
```
Comment (kestarumper) [inline packages/uniforms/src/ValidatedForm.tsx:135]: Why is this `throw`ed here if we follow up with `.catch(noop)` later on (line 139)?

File: packages/uniforms/__tests__/ValidatedForm.tsx (hunk 1 above)
Comment (radekmie) [reply packages/uniforms/__tests__/ValidatedForm.tsx:391]: I'll extract it to a separate function with an appropriate name - it should help.

File: packages/uniforms/src/ValidatedForm.tsx (hunk 2 above)
Comment (radekmie) [reply packages/uniforms/src/ValidatedForm.tsx:135]: Think of it as of a _re-rejection_ (as in _re-throwing_ an error). Not doing that would effectively swallow `onSubmit` errors.

The `result.catch(noop);` below is there just to suppress the `Unhandled rejection` errors. Every validation error leads to a rejected promise (that `Promise.reject` in line 128) but it is _technically_ handled (i.e. it's already in the state).

File: packages/uniforms/src/ValidatedForm.tsx [hunk 3]
```diff
@@ -171,41 +150,54 @@ export function Validated<Base extends typeof BaseForm>(Base: Base) {
...
+    if (value instanceof Promise) {
+      callIfAsync();
+      return value.then(fn);
+    }
+
+    return fn(value);
+  }
+
+  return then;
+}
```
Comment (Monteth) [inline packages/uniforms/src/ValidatedForm.tsx:199]: That's brilliant 👑

**Pull request merged**